src/Arcus.WebApi.Unit/Security/Authentication/CertificateAuthenticationFilterTests.cs

```diff
@@ -0,0 +1,129 @@
+                        NoneAuthenticationController.Route);
+
+                    // Act
+                    using (HttpResponseMessage response = await client.SendAsync(request))
+                    {
+                        // Arrange
+                        Assert.Equal(HttpStatusCode.Unauthorized, response.StatusCode);
+                    }
+                }
+            }
+        }
+
+        [Theory]
+        [InlineData("", false)]
+        [InlineData("thumbprint-noise", true)]
+        public async Task AuthorizedRoute_WithCertificateAuthentication_ShouldFailWithUnauthorized_WhenClientCertificateThumbprintDoesntMatch(
+            string thumbprintNoise,
+            bool expected)
+        {
+            // Arrange
+            using (X509Certificate2 clientCertificate = SelfSignedCertificate.Create())
+            {
+                const string thumbprintKey = "thumbprint";
+                _testServer.AddService<ISecretProvider>(new InMemorySecretProvider((thumbprintKey, clientCertificate.Thumbprint + thumbprintNoise)));
+                _testServer.AddFilter(new CertificateAuthenticationFilter(X509ValidationRequirement.Thumbprint, thumbprintKey));
+                _testServer.SetClientCertificate(clientCertificate);
+
+                using (HttpClient client = _testServer.CreateClient())
+                {
+                    var request = new HttpRequestMessage(
+                        HttpMethod.Get,
+                        NoneAuthenticationController.Route);
+
+                    // Act
+                    using (HttpResponseMessage response = await client.SendAsync(request))
+                    {
+                        // Arrange
+                        Assert.True(
+                            (HttpStatusCode.Unauthorized == response.StatusCode) == expected,
+                            $"Response HTTP status code {(expected ? "should" : "shouldn't")} be 'Unauthorized' but was '{response.StatusCode}'");
+                    }
+                }
+            }
+        }
+
+        [Theory]
+        [InlineData("known-subject", "known-issuername", false)]
+        [InlineData("unrecognizedSubjectName", "known-issuername", true)]
+        [InlineData("known-subject", "unrecognizedIssuerName", true)]
+        [InlineData("unrecognizedSubjectName", "unrecognizedIssuerName", true)]
+        public async Task AuthorizedRoute_WithCertificateAuthentication_ShouldFailWithUnauthorized_WhenAnyClientCertificateValidationDoesntSucceeds(
+            string subjectValue,
+            string issuerValue,
+            bool expected)
+        {
+            // Arrange
+            const string subjectKey = "subject", issuerKey = "issuer";
+            _testServer.AddService<ISecretProvider>(
+                new InMemorySecretProvider(
+                    (subjectKey, "CN=known-subject"),
+                    (issuerKey, "CN=known-issuername")));
+
+            _testServer.AddFilter(
```

Comment:
This notation feels very unnatural for object-oriented programming, why are we not using a list or object with all requirements inside of it?

For example:
```csharp
var certificateRequirements = new CertificateRequirements
{
    SubjectNames = new List<string>() {subjectKey},
    IssuerNames= new List<string>() {issuerKey},
};

_testServer.AddFilter(new CertificateAuthenticationFilter(certificateRequirements));
```

What do you think @fgheysels

Reply:
Based on reading the code it looks like `X509ValidationRequirement.SubjectName` is a constant string but if I have a look at the constructor this is actually a requirement or?

The code is not clear on that perspective imo, sorry :/

Reply:
I wouldn't use a `List` type as a parameter for the constructor of the `CertificateRequirements` class since it is to concrete and has functionality that is not necessary.
The code in the PR is possible because of the `params` keyword in the constructor.  It allows you to write an array as a comma separated list of values.  Given the fact that the type of the array is a ValueTuple, the syntax `( ... ) , ( ... )` is possible.
It is true that it can be somewhat confusing when you read it.

Therefore, I would suggest to change the constructor of the `CertificateAuthenticationFilter` to:

```csharp
public CertificateAuthenticationFilter( IEnumerable<(X509ValidationRequirement, string)> )
```
So, using an `IEnumerable` instead of the more concrete/specific array type.

This means that the code in question needs to be written like this:

```csharp
new CertificateAuthenticationFilter( new [] 
                                     { 
                                         (X509ValidationRequirement.SubjectName, subjectKey),
                                         (X509ValidationRequirement.IssuerName, issuerKey)
                                     });
```

Reply:
See most latest comment on new builder approach.

Resolving this in favor of the other one.